### PR TITLE
add a special API to get a whitelist bitvector

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst @toregge please review

### CODE REVIEW

**Comment by arnej27959** [comment]
one thing i don't know: do we need to use "invalidateCachedCount" on the bitvector?  because NN blueprint uses countTrueBits.

File: searchcore/src/vespa/searchcore/proton/documentmetastore/white_list_provider.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/searchlib/common/bitvector.h>
+
+#pragma once
+
+namespace proton::documentmetastore {
+
+struct WhiteListProvider {
+    virtual search::BitVector::UP get_white_list_filter() const = 0;
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/white_list_provider.h:10]: Consider forward declaring search::BitVector and using std::unique_ptr<search::BitVector>.

File: searchcore/src/vespa/searchcore/proton/matching/query.cpp [hunk 2]
```diff
@@ -156,6 +157,11 @@ void
 Query::setWhiteListBlueprint(Blueprint::UP whiteListBlueprint)
 {
     _whiteListBlueprint = std::move(whiteListBlueprint);
+    using proton::documentmetastore::WhiteListProvider;
+    auto wlf = dynamic_cast<WhiteListProvider *>(_whiteListBlueprint.get());
+    if (wlf) {
+        _global_white_list = wlf->get_white_list_filter();
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/matching/query.cpp:163]: This adds a cost (creating a copy of the white list bitvector) even when this copy won't be used.

**Review by toregge — CHANGES_REQUESTED** [review]

File: searchcore/src/vespa/searchcore/proton/matching/query.cpp (hunk 2 above)
Comment (geirst) [reply searchcore/src/vespa/searchcore/proton/matching/query.cpp:163]: Agree with Tor. Can we do this in optimize() instead, where we know we need it? In that case we need to store a pointer to WhiteListProvider in Query as the white list blueprint are moved into the blueprint tree between this call and the call to optimize().

**Review by geirst — COMMENTED** [review]
We need to find a solution to avoid creating the global white list when we don't need it, otherwise LGTM.

File: searchcore/src/vespa/searchcore/proton/matching/query.h [hunk 3]
```diff
@@ -19,11 +21,12 @@ class Query
 {
 private:
     using Blueprint=search::queryeval::Blueprint;
+    using WhiteListProvider=proton::documentmetastore::WhiteListProvider;
     search::query::Node::UP _query_tree;
     Blueprint::UP           _blueprint;
     search::fef::Location   _location;
     Blueprint::UP           _whiteListBlueprint;
-    search::BitVector::UP   _global_white_list;
+    WhiteListProvider      *_white_list_provider;
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/matching/query.h:29]: This could be const as get_white_list_filter() is const.

**Review by geirst — APPROVED** [review]
👍

**Review by toregge — APPROVED** [review]

**Pull request merged**